### ISSUE TITLE
Gsettings Tracker

### ISSUE THREAD

**Issue by mpranj** [ElektraInitiative/libelektra#3403]
This issue tracks the state of Elektra's gsettings backend for documentation purposes. We will decide later which parts need a separate issue.

Needs fixing:

- [ ] ~~if no `.config/default.ecf` exists none is created, so any configuration updates are lost.~~ I think the statement is no longer true, but will re-check it.

**Comment by markus2330** [comment]
Thank you for creating the issue!

>  if no .config/default.ecf exists none is created, so any configuration updates are lost.

Weird, kdbSet should always create it with non-empty configs?

**Comment by mpranj** [comment]
> Weird, kdbSet should always create it with non-empty configs?

Maybe it did not even reach kdbSet. I will just document anything suspicious as it happens and re-check it later.

**Comment by mpranj** [comment]
Quick update: 

It seems that GSettings uses our (current) ElektraGSettings backend in an inappropriate way. I saw random crashes and weird behaviour until I noticed in a trace that multiple threads seemingly access the same KDB handle and KeySets concurrently. Unfortunately those are not thread-safe.

Wrapping critical accesses with a global lock seems to eliminate the random crashes. I'll have to investigate how this impacts performance and if it can be improved.

Ideally the threads would have separate KDB handles, but upstream GSettings seems to simply clone the ElektraGSettings instance (thus also the KDB handle) after initialization. So if I am not mistaken, we will have to provide a thread-safe implementation from our side.

**Comment by markus2330** [comment]
Thanks for the update!

Is there no API entry/hook/... when ElektraGSettings is cloned? Then we could create a new KDB handle on every clone.

There is every possibility, however, that the lock is faster: Depending on how often it gets cloned and how much concurrent access occurs.

Could be an interesting benchmark to compare these two methods.

**Comment by mpranj** [comment]
As noted in the last meeting the GNOME team asked me to extract the concurrent access problem described above to a minimal example.
 
I was unable to reproduce the problem now and seems like it was fixed in the mean time. Thus I removed the locks and everything is still working fine. :+1:

Also as already mentioned, the separate `dbus` worker is _probably_ not needed, so I decided we will not implement it unless it becomes non-optional.

Some small fixes are needed before I update my PR: reading default keys (i.e. from `system` namespace) is not working / not properly implemented. Otherwise, the `user` namespace works quite well afaict.
